tc/autotuner/options_cache.h

```diff
@@ -0,0 +1,173 @@
+  using MultiMapType = std::unordered_multimap<
+      OptionsCacheKey,
+      OptionsCacheValue<Backend>,
+      OptionsCacheKeyHash>;
+
+  OptionsCache<Backend>() = default;
+  OptionsCache<Backend>(const OptionsCache<Backend>& other);
+
+  /// Clears the content of the cache and resets the counters
+  void clear();
+
+  /// \return the number of values for a particular key
+  size_t count(const OptionsCacheKey& key) const;
+
+  //// \return the number of elements in the cache
+  size_t size() const;
+
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromFile(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void storeCacheToFile(const std::string& filename) const;
+
+  /// Saves a new runtime.
+  /// If the key does not exist, a new entry is inserted.
+  /// If the key exists, a search is performed on options.
+  /// If the corresponding options are found, the duration is appended to the
+  /// runtimes, otherwise a new entry is inserted in the multimap.
+  void recordRuntime(
+      const lang::CanonicalTcString& tc,
+      const std::vector<TensorInfo>& inputs,
+      const std::vector<TensorInfo>& outputs,
+      const std::string& deviceStr,
+      const typename Backend::MappingOptionsType& options,
+      Duration duration);
+
+  /// Returns the top-K best mapping options for a particular
+  /// TC/inputs/outputs/device. Note that the result may be empty (in
+  /// particular if problem size is small and pruning threshold is too high
+  /// for the problem size).
+  /// \returns a vector of mapping options
+  std::vector<typename Backend::MappingOptionsType> getTopKOptions(
+      const lang::CanonicalTcString& tc,
+      const std::vector<TensorInfo>& inputs,
+      const std::vector<TensorInfo>& outputs,
+      const std::string& deviceStr,
+      size_t K) const;
+
+  /// Drops the (N - K) worst performing options
+  void pruneKeepTopK(size_t K);
+
+ protected:
+  // Make protected and not private so we can derive and test the internals
+  typename Backend::OptionsCacheProtoType toProtobuf() const;
+  void fromProtobuf(const typename Backend::OptionsCacheProtoType& proto);
+
+ public:
+  mutable size_t numberCacheAttempts{0};
+  mutable size_t numberAttemptedRetrievals{0};
+  mutable size_t numberSuccessfulRetrievals{0};
+
+ protected:
+  // Make protected and not private so we can derive and test the internals
+  mutable std::mutex mutex;
+
+  std::unordered_multimap<
+      OptionsCacheKey,
+      OptionsCacheValue<Backend>,
+      OptionsCacheKeyHash>
+      store_;
+};
+} // namespace autotune
+
+inline std::string makeOptionsFilename(const std::string& fn) {
+  return fn + ".options";
+}
```

Comment:
I'm not sure about the exact rule for putting things in `-inl.h`, but perhaps
this should be put there too.